### PR TITLE
lib/libc: do not enable `picolibc` if neither the toolchain has it nor its module is present

### LINKED ISSUE #75286: OOT board fails to build after going from v3.5.0 to v3.7.0
**Describe the bug**
I encountered the following compilation error while testing build on v3.7.0 for OOT boards.

```
are/zephyr/__sources-3.7.0__/out/zephyr/subsys/debug/symtab/symtab_stub.c
riscv64-unknown-elf-gcc: fatal error: cannot read spec file 'picolibc.specs': No such file or directory
compilation terminated.
ninja: build stopped: subcommand failed.
```

This didn't happen in v3.5.0.

There's a similar report in the `#cpp` channel in our discord server, but I'm not sure what board they used.

https://discord.com/channels/720317445772017664/883444430676324382/1255783684322230282

Screenshot 2024-07-02 at 12 17 29 AM

Report from the user: 
> ```
> NCS - HEAD detached at v2.7.0-rc3
> Zephyr - HEAD detached at 3e80418dd59
> ```
>
> WSL2
> ```
> Distributor ID: Ubuntu
> Description:    Ubuntu 22.04.4 LTS
> Release:        22.04
> Codename:       jammy
> ```
>
> CMake
> ```
> -DBOARD=native_sim
> ```
> 
> ```
> - CMake version: 3.28.1
> -- Cache files will be written to: /home/dev/.cache/zephyr
> -- Zephyr version: 3.6.99 (/home/dev/ncs/zephyr)
> -- Found west (found suitable version "1.2.0", minimum required is "0.14.0")
> -- Board: native_sim, qualifiers: native
> -- Found host-tools: zephyr 0.16.8 (/home/dev/zephyr-sdk-0.16.8)
> -- Found toolchain: host (gcc/ld)
> ```
> 
> ```
> CONFIG_CPP=y
> CONFIG_GLIBCXX_LIBCPP=y
> CONFIG_POSIX_API=y
> CONFIG_PICOLIBC=y
> ```

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using?
     - OOT RISC-V boards with OOT SoCs
 - What have you tried to diagnose or workaround this issue? 
     - Using `MINIMAL_LIBC` can make the error to go away
 - Is this a regression? If yes, have you been able to "git bisect" it to a
   specific commit?
     - Kinda? It used to build in v3.5

**To Reproduce**
Probably not easy to reproduce with upstream SDK.

**Expected behavior**
I'm expecting my application to build on v3.7.0

**Impact**
OOT board/soc/application doesn't build on v3.7.0

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Custom gcc toolchain
 - Commit SHA or Version used: 3.7.0-rc1

**Root cause**
The cause of the issue that I've encountered is due to PICOLIBC getting selected

https://github.com/zephyrproject-rtos/zephyr/blob/cac70373aa81b9989905ac530b9773e4b748e409/lib/libc/Kconfig#L60-L65

instead of yielding to the next default even though the build env doesn't have the source for it (OOT application may not be using Zephyr SDK + the PICOLIBC module).

This issue can be worked-around for users going from 2.7 LTS to 3.7 LTS by **explicitly** selecting the minimal libc (`CONFIG_MINIMAL_LIBC=y`)

**Log**

```
stdout:
[1/215] Preparing syscall dependency handling

[2/215] Generating include/generated/zephyr/version.h
-- Zephyr version: 3.7.0-rc1 (/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr), build: 3.7.0-fb
[3/215] Generating misc/generated/syscalls_subdirs.trigger
[4/215] Generating misc/generated/syscalls.json, misc/generated/struct_tags.json
[5/215] Generating include/generated/zephyr/syscall_dispatch.c, include/generated/zephyr/syscall_list.h
[6/215] Generating include/generated/zephyr/driver-validation.h
[7/215] Generating include/generated/zephyr/kobj-types-enum.h, include/generated/zephyr/otype-to-str.h, include/generated/zephyr/otype-to-size.h
[8/215] Building C object zephyr/CMakeFiles/offsets.dir/arch/riscv/core/offsets/offsets.c.obj
FAILED: zephyr/CMakeFiles/offsets.dir/arch/riscv/core/offsets/offsets.c.obj 
/mnt/infra-asic-riscv64-toolchain-20230823/bin/riscv64-unknown-elf-gcc -DKERNEL -DK_HEAP_MEM_POOL_SIZE=10240 -DPICOLIBC_DOUBLE_PRINTF_SCANF -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR__=1 -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/kernel/include -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/arch/riscv/include -Izephyr/include/generated/zephyr -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/include -Izephyr/include/generated -I/data/users/ycsin/firmware/zephyr/modules/meta_asic_latest/__meta_asic_latest__/meta_asic_latest/soc/meta/mtia -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/include/zephyr/posix -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/lib/posix/options/getopt -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/soc/common/riscv-privileged/. -I/data/users/ycsin/firmware/zephyr/modules/meta_asic_latest/__meta_asic_latest__/meta_asic_latest/soc/meta/mtia/riscv_board/. -I/data/users/ycsin/firmware/zephyr/modules/meta_asic_latest/__meta_asic_latest__/meta_asic_latest/soc/meta/mtia/riscv_board/../meta-common -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/subsys/tracing/include -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/subsys/tracing/user/. -I/data/users/ycsin/firmware/zephyr/modules/meta_asic_latest/__meta_asic_latest__/meta_asic_latest/include -I/data/users/ycsin/firmware/zephyr/modules/meta_asic_latest/__meta_asic_latest__/meta_asic_latest/lib/benchmark/sub_cmds/.. -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/subsys/testsuite/include -I/data/users/ycsin/firmware/zephyr/modules/hal/mtia_gen/__mtia_gen__/out/mtia_gen/include -isystem /data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/lib/libc/common/include -fno-strict-aliasing -Os -imacros /data/users/ycsin/firmware/zephyr/__riscv_board-latest-cmake__/out/facebook/configs/riscv_board/zephyr/include/generated/zephyr/autoconf.h -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcmodel=medany -mabi=lp64d -march=rv64imafdc_zicsr_zifencei --sysroot=/mnt/infra-asic-riscv64-toolchain-20230823/riscv64-unknown-elf -imacros /data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/data/users/ycsin/firmware/zephyr/modules/meta_asic_latest/__meta_asic_latest__/meta_asic_latest/facebook/configs/riscv_board=CMAKE_SOURCE_DIR -fmacro-prefix-map=/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr=ZEPHYR_BASE -ffunction-sections -fdata-sections --specs=picolibc.specs -D_POSIX_THREADS -mexplicit-relocs -std=c99 -fno-lto -MD -MT zephyr/CMakeFiles/offsets.dir/arch/riscv/core/offsets/offsets.c.obj -MF zephyr/CMakeFiles/offsets.dir/arch/riscv/core/offsets/offsets.c.obj.d -o zephyr/CMakeFiles/offsets.dir/arch/riscv/core/offsets/offsets.c.obj -c /data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/arch/riscv/core/offsets/offsets.c
riscv64-unknown-elf-gcc: fatal error: cannot read spec file 'picolibc.specs': No such file or directory
compilation terminated.
[9/215] Building C object zephyr/subsys/debug/symtab/CMakeFiles/symtab.dir/symtab_stub.c.obj
FAILED: zephyr/subsys/debug/symtab/CMakeFiles/symtab.dir/symtab_stub.c.obj 
/mnt/infra-asic-riscv64-toolchain-20230823/bin/riscv64-unknown-elf-gcc -DKERNEL -DK_HEAP_MEM_POOL_SIZE=10240 -DPICOLIBC_DOUBLE_PRINTF_SCANF -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR__=1 -Izephyr/include/generated/zephyr -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/include -Izephyr/include/generated -I/data/users/ycsin/firmware/zephyr/modules/meta_asic_latest/__meta_asic_latest__/meta_asic_latest/soc/meta/mtia -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/include/zephyr/posix -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/lib/posix/options/getopt -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/soc/common/riscv-privileged/. -I/data/users/ycsin/firmware/zephyr/modules/meta_asic_latest/__meta_asic_latest__/meta_asic_latest/soc/meta/mtia/riscv_board/. -I/data/users/ycsin/firmware/zephyr/modules/meta_asic_latest/__meta_asic_latest__/meta_asic_latest/soc/meta/mtia/riscv_board/../meta-common -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/arch/riscv/include -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/kernel/include -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/subsys/tracing/include -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/subsys/tracing/user/. -I/data/users/ycsin/firmware/zephyr/modules/meta_asic_latest/__meta_asic_latest__/meta_asic_latest/include -I/data/users/ycsin/firmware/zephyr/modules/meta_asic_latest/__meta_asic_latest__/meta_asic_latest/lib/benchmark/sub_cmds/.. -I/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/subsys/testsuite/include -I/data/users/ycsin/firmware/zephyr/modules/hal/mtia_gen/__mtia_gen__/out/mtia_gen/include -isystem /data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/lib/libc/common/include -Wshadow -fno-strict-aliasing -Os -imacros /data/users/ycsin/firmware/zephyr/__riscv_board-latest-cmake__/out/facebook/configs/riscv_board/zephyr/include/generated/zephyr/autoconf.h -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcmodel=medany -mabi=lp64d -march=rv64imafdc_zicsr_zifencei --sysroot=/mnt/infra-asic-riscv64-toolchain-20230823/riscv64-unknown-elf -imacros /data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/data/users/ycsin/firmware/zephyr/modules/meta_asic_latest/__meta_asic_latest__/meta_asic_latest/facebook/configs/riscv_board=CMAKE_SOURCE_DIR -fmacro-prefix-map=/data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr=ZEPHYR_BASE -ffunction-sections -fdata-sections --specs=picolibc.specs -D_POSIX_THREADS -mexplicit-relocs -std=c99 -MD -MT zephyr/subsys/debug/symtab/CMakeFiles/symtab.dir/symtab_stub.c.obj -MF zephyr/subsys/debug/symtab/CMakeFiles/symtab.dir/symtab_stub.c.obj.d -o zephyr/subsys/debug/symtab/CMakeFiles/symtab.dir/symtab_stub.c.obj -c /data/users/ycsin/firmware/zephyr/__sources-3.7.0__/out/zephyr/subsys/debug/symtab/symtab_stub.c
riscv64-unknown-elf-gcc: fatal error: cannot read spec file 'picolibc.specs': No such file or directory
compilation terminated.
ninja: build stopped: subcommand failed.
```

### PR DESCRIPTION
**Pull request by ycsin**
Currently, `PICOLIBC` is getting enabled even if neither the toolchain has it nor the picolibc module is available to the build system. This can be the case when Zephyr is built in OOT environment, with external toolchain and without `West`.

A possible workaround to the issue is to specify another LIBC (i.e. `CONFIG_MINIMAL_LIBC=y`) explicitly, but in any case, we probably do not want to compile the PICOLIBC interface when we do not have its source.

Fixes #75286

### CODE REVIEW

**Comment by henrikbrixandersen** [comment]
CC: @keith-packard

File: lib/libc/Kconfig [hunk 1]
```diff
@@ -43,7 +43,8 @@ config NEWLIB_LIBC_SUPPORTED
 config PICOLIBC_SUPPORTED
 	bool
 	depends on !NATIVE_APPLICATION
-	depends on !REQUIRES_FULL_LIBCPP || ("$(TOOLCHAIN_HAS_PICOLIBC)" = "y")
+	depends on !REQUIRES_FULL_LIBCPP
```
Comment (keith-packard) [inline lib/libc/Kconfig:46]: This looks wrong -- the logic here is that you must use the toolchain picolibc when using the C++ library, but it looks like you've made picolibc completely incompatible with the C++ library.

File: lib/libc/picolibc/Kconfig [hunk 2]
```diff
@@ -16,7 +16,7 @@ config PICOLIBC_USE_MODULE
 # force TLS when using toolchain with TLS support
 config PICOLIBC_USE_TOOLCHAIN
 	bool
-	default y if PICOLIBC && !PICOLIBC_USE_MODULE
```
Comment (keith-packard) [inline lib/libc/picolibc/Kconfig:19]: I don't think this change is useful, and probably not correct. You only want to use the toolchain picolibc when you're using picolibc.

**Review by keith-packard — CHANGES_REQUESTED** [review]
Need to fix the logic to continue allowing use of picolibc with the full C++ library.

File: lib/libc/picolibc/Kconfig (hunk 2 above)
Comment (ycsin) [reply lib/libc/picolibc/Kconfig:19]: the `PICOLIBC` here should be redundant:

https://github.com/zephyrproject-rtos/zephyr/blob/e912a95355580ed3ae635f02103a6c7a3859611e/lib/libc/picolibc/Kconfig#L4

Comment (keith-packard) [reply lib/libc/picolibc/Kconfig:19]: ah, great point. Thanks for providing the wider context. Reviewing patches is always tricky.

File: lib/libc/Kconfig (hunk 1 above)
Comment (ycsin) [reply lib/libc/Kconfig]: probably fixed, could you give this another spin?

**Comment by keith-packard** [comment]
When we discussed picolibc integration a while ago I recall a general consensus that it was OK to depend on having the picolibc module available during the build. Would it be better to detect that neither module nor toolchain library were present and report an error?

**Comment by ycsin** [comment]
> Would it be better to detect that neither module nor toolchain library were present and report an error?

Do you mean that the build should error (and fail) if neither module nor toolchain library were present?

**Comment by keith-packard** [comment]
> Do you mean that the build should error (and fail) if neither module nor toolchain library were present?

I was thinking so, but now I'm not sure. I guess I wonder why you don't want to have the modules present for this? A simple 'west update' and you get a better C library. We've had a couple of people talk about the minimal C library as 'sufficient for validating core Zephyr code', but not really the best choice for application developers. Maybe we want to require developers to explicitly ask for the minimal C library so they aren't surprised by it's limitations when they fail to set up the system completely?

So, I think it comes down to whether you think it's reasonable to want Zephyr to build out of the box without having the modules present. I'm hoping to see the minimal C library de-emphasized even further as it's an unnecessary drain on Zephyr developers; we don't need another embedded C library in 2024.

**Comment by aescolar** [comment]
> I guess I wonder why you don't want to have the modules present for this? A simple 'west update' and you get a better C library. 

@keith-packard note this line in the PR description:

> This can be the case when Zephyr is built in OOT environment, with external toolchain and without West

Independently of Picolibc being a nicer C library than quite a few others, we support users using their own toolchain and C library. They may need to or prefer using the C library that comes with their tool chain or which is optimized for their target.

**Comment by ycsin** [comment]
> I guess I wonder why you don't want to have the modules present for this? A simple 'west update' and you get a better C library.

I dearly wish that I could, but my company is using [buck2](https://buck2.build/) instead of CMake & [Mercurial](https://www.mercurial-scm.org/) instead of git, that combination basically ruled out `west` unfortunately.

> So, I think it comes down to whether you think it's reasonable to want Zephyr to build out of the box without having the modules present. I'm hoping to see the minimal C library de-emphasized even further as it's an unnecessary drain on Zephyr developers; we don't need another embedded C library in 2024.

My view is that modules should be modular. Unless picolibc is baked directly into the Zephyr repo, I probably do not want to remove the minimal libc, but I guess that decision isn't up to me to make.

This PR is just trying to fix an issue where picolibc is enabled when it shouldn't, it should be a no-op for (most) users that are fortunate enough to be able to use `west` & the Zephyr SDK, picolibc continues to be the default, but fallback to something else when it isn't available.

**Comment by cfriedt** [comment]
> > Do you mean that the build should error (and fail) if neither module nor toolchain library were present?
> 
> I was thinking so, but now I'm not sure. 

It probably does make sense for a build to error and fail if the required components are not present.

That's what happens in most of the cases that I can think of.

**Comment by keith-packard** [comment]
> Independently of Picolibc being a nicer C library than quite a few others, we support users using their own toolchain and C library. They may need to or prefer using the C library that comes with their tool chain or which is optimized for their target.

Absolutely -- a toolchain supplied C library is probably the best option for most people. The issue here was that the toolchain didn't supply a C library and the picolibc module was unavailable. In that case, we have a choice of either failing the build or resorting to the minimal C library.

**Comment by keith-packard** [comment]
> I dearly wish that I could, but my company is using [buck2](https://buck2.build/) instead of CMake & [Mercurial](https://www.mercurial-scm.org/) instead of git, that combination basically ruled out `west` unfortunately.

If you're interested in using picolibc for this environment, then perhaps you could integrate it into your homebrew toolchain?

> My view is that modules should be modular. Unless picolibc is baked directly into the Zephyr repo, I probably do not want to remove the minimal libc, but I guess that decision isn't up to me to make.

A few core Zephyr developers have asked that the minimal C library be kept around as a way of doing kernel development and validation without needing anything outside of the Zephyr repository. So it's likely to hang around in some form at least for a while.

> This PR is just trying to fix an issue where picolibc is enabled when it shouldn't, it should be a no-op for (most) users that are fortunate enough to be able to use `west` & the Zephyr SDK, picolibc continues to be the default, but fallback to something else when it isn't available.

I assume you can manually configure the minimal C library and your project builds successfully? If so, that leaves us needing to answer the question of whether Zephyr should automatically switch to the minimal C library or emit an error if no other C library is available (either a toolchain C library or the picolibc module).

As you say, whatever is done here shouldn't affect other users, so it seems like you are probably the best person to answer this.

**Comment by ycsin** [comment]
> I assume you can manually configure the minimal C library and your project builds successfully? If so, that leaves us needing to answer the question of whether Zephyr should automatically switch to the minimal C library or emit an error if no other C library is available (either a toolchain C library or the picolibc module).

Yeeeeah, I managed to get around this by specifying the minimal libc explicitly in my applications, but I shouldn't have to do so as the `choice LIBC_IMPLEMENTATION` should have taken care of that for me by automatically switching to the next default:

https://github.com/zephyrproject-rtos/zephyr/blob/cac70373aa81b9989905ac530b9773e4b748e409/lib/libc/Kconfig#L60-L65

that fact that it didn't means that something isn't working as expected IMHO, and this PR is meant to plug that hole and make everyone's transition to the v3.7.0 more seamless.

File: lib/libc/Kconfig [hunk 3]
```diff
@@ -43,7 +43,7 @@ config NEWLIB_LIBC_SUPPORTED
 config PICOLIBC_SUPPORTED
 	bool
 	depends on !NATIVE_APPLICATION
-	depends on !REQUIRES_FULL_LIBCPP || ("$(TOOLCHAIN_HAS_PICOLIBC)" = "y")
```
Comment (keith-packard) [inline lib/libc/Kconfig:46]: Let's figure out what this should look like.

For apps using the full C++ library, `PICOLIBC_SUPPORTED` depends on `TOOLCHAIN_HAS_PICOLIBC`. No other option is supported.

For all other apps, `PICOLIBC_SUPPORTED` depends on *either* `TOOLCHAIN_HAS_PICOLIBC` or `ZEPHYR_PICOLIBC_MODULE`.

Let's just make that right; I don't think we need to change anything other than this; once `PICOLIBC_SUPPORTED` is false, the existing libc selection logic will fall through to the minimal C library.

Here's the full logic table:
```
libc++  TOOLCHAIN       MODULE          PICOLIBC_SUPPORTED
  0         0             0                     0
  0         0             1                     1
  0         1             0                     1
  0         1             1                     1
  1         0             0                     0
  1         0             1                     0
  1         1             0                     1
  1         1             1                     1
```
By examination, we can see that
```
PICOLIBC_SUPPORTED = TOOLCHAIN || (MODULE && !libc++)
```
So, we can change the above depends line to:
```
depends on ("$(TOOLCHAIN_HAS_PICOLIBC)" = "y") || (ZEPHYR_PICOLIBC_MODULE && !REQUIRES_FULL_LIBCPP)
```
Feel free to flip the disjunction around to make it look more like the original code.

Comment (ycsin) [reply lib/libc/Kconfig:46]: Updated the commit (&message)

**Review by henrikbrixandersen — DISMISSED** [review]

**Review by keith-packard — COMMENTED** [review]
I don't think the changes to picolibc/Kconfig are required to solve this problem; perhaps I just don't understand the full problem?

**Review by cfriedt — DISMISSED** [review]

**Comment by ycsin** [comment]
> I don't think the changes to picolibc/Kconfig are required to solve this problem; perhaps I just don't understand the full problem?

Yup, they are not required anymore after fixing the `PICOLIBC_SUPPORTED`, I've spinned off the 3rd commit to #75483 as it is not a bug fix

**Review by keith-packard — APPROVED** [review]
looking good.

**Review by henrikbrixandersen — APPROVED** [review]

**Pull request merged**